### PR TITLE
Print stack trace

### PR DESCRIPTION
**Pull request by rikhuijzer**
When using Franklin, I usually know where the error is coming from, so then the current warning which only shows the error type is fine. However, at times I duly miss the stacktrace and am completely in the dark about the cause of an error.

Currently, this PR proposes to print the stacktrace to `stderr`. I think this is more convenient for the user than storing it in `outf`. (I also don't think the `outf` has ever been of much help for me. It typically contains the error I already had in the REPL in one file, and `nothing` in the other.)

The response is now something like 
```
DomainError with -1.0:
sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(::Symbol, ::Float64) at ./math.jl:33
 [2] sqrt at ./math.jl:573 [inlined]
 [3] sqrt(::Int64) at ./math.jl:599
 [4] top-level scope at none:1
 [5] eval at ./boot.jl:331 [inlined]
 [6] (::Franklin.var"#96#98"{Module,Array{Any,1},Int64})() at /home/rik/FR/src/eval/run.jl:64
 [7] redirect_stdout(::Franklin.var"#96#98"{Module,Array{Any,1},Int64}, ::IOStream) at ./stream.jl:1150
 [8] (::Franklin.var"#95#97"{Module,String,Array{Any,1},Int64})(::IOStream) at /home/rik/FR/src/eval/run.jl:60
 [9] open(::Franklin.var"#95#97"{Module,String,Array{Any,1},Int64}, ::String, ::Vararg{String,N} where N; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at ./io.jl:325
 [10] open at ./io.jl:323 [inlined]
 [11] run_code(::Module, ::String, ::String; warn_err::Bool, strip_code::Bool) at /home/rik/FR/src/eval/run.jl:56
 [12] run_code(::Module, ::String, ::String) at /home/rik/FR/src/eval/run.jl:49
 [13] macro expansion at /home/rik/FR/test/eval/run.jl:78 [inlined]
 [14] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115 [inlined]
 [15] top-level scope at /home/rik/FR/test/eval/run.jl:26
 [16] include(::String) at ./client.jl:457
 [17] top-level scope at /home/rik/FR/test/runtests.jl:50
 [18] include(::String) at ./client.jl:457
 [19] top-level scope at none:6
 [20] eval(::Module, ::Any) at ./boot.jl:331
 [21] exec_options(::Base.JLOptions) at ./client.jl:272
 [22] _start() at ./client.jl:506
┌ Franklin Warning: in <input string>
│ There was an error of type 'DomainError' when running a code block.
│ Checking the output files '/tmp/jl_yNUxkV.(out|res)'
│ might be helpful to understand and solve the issue.
│
│ Relevant pointers:
│ - evaluation of Julia code blocks: https://franklinjl.org/code/
└
```

The code is not ready to be merged yet, it needs more testing and possibly a bit of tweaking of the design.

### CODE REVIEW

**Comment by tlienart** [comment]
Thanks for  this Rik! This seems useful though I think in terms of legibility, it might make more sense to reproduce the top part of the  stack trace either in or below the warning rather than above (you can ignore lines further than `run_code` btw) what do  you think?

**Comment by rikhuijzer** [comment]
Good point about users trying to fix multiple errors! I didn't think of that.

I agree that it is better inside the warning. Now, the output is
```
┌ Franklin Warning: in <input string>
│ There was an error of type 'DomainError' when running a code block.
│ Checking the output files '/tmp/jl_lfP2Yh.(out|res)'
│ might be helpful to understand and solve the issue.
│
│ Relevant pointers:
│ - evaluation of Julia code blocks: https://franklinjl.org/code/
│
│ Stacktrace:
│ DomainError with -1.0:
│ sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
│ Stacktrace:
│  [1] throw_complex_domainerror(::Symbol, ::Float64) at ./math.jl:33
│  [2] sqrt at ./math.jl:573 [inlined]
│  [3] sqrt(::Int64) at ./math.jl:599
│  [4] top-level scope at none:1
│  [5] eval at ./boot.jl:331 [inlined]
│  [6] (::Franklin.var"#96#98"{Module,Array{Any,1},Int64})() at /home/rik/git/FR/src/eval/run.jl:65
│  [7] redirect_stdout(::Franklin.var"#96#98"{Module,Array{Any,1},Int64}, ::IOStream) at ./stream.jl:1150
│  [8] (::Franklin.var"#95#97"{Module,String,Array{Any,1},Int64})(::IOStream) at /home/rik/git/FR/src/eval/run.jl:61
│  [9] open(::Franklin.var"#95#97"{Module,String,Array{Any,1},Int64}, ::String, ::Vararg{String,N} where N; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at ./io.jl:325
│  [10] open at ./io.jl:323 [inlined]
│  [11] run_code(::Module, ::String, ::String; warn_err::Bool, strip_code::Bool) at /home/rik/git/FR/src/eval/run.jl:57
│  [12] run_code(::Module, ::String, ::String) at /home/rik/git/FR/src/eval/run.jl:49
│  [13] macro expansion at /home/rik/git/FR/test/eval/run.jl:79 [inlined]
│  [14] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115 [inlined]
│  [15] top-level scope at /home/rik/git/FR/test/eval/run.jl:26
│  [16] include(::String) at ./client.jl:457
│  [17] top-level scope at /home/rik/git/FR/test/runtests.jl:50
│  [18] include(::String) at ./client.jl:457
│  [19] top-level scope at none:6
│  [20] eval(::Module, ::Any) at ./boot.jl:331
│  [21] exec_options(::Base.JLOptions) at ./client.jl:272
│  [22] _start() at ./client.jl:506
└
```

I put the stacktrace at the bottom, because more experienced users would be mostly interested in the trace. New users can now still scroll up and see the more friendly warning message. 

I also took the liberty of cleaning up and removing the TODO. Let me know if I should put it back.

**Comment by tlienart** [comment]
ok yes that's great! maybe one last small thing; the stack trace here is only useful up to line `[5]` ; you should be able to regex the next line which should be the same always:

```
[6] (::Franklin.var"#96#98"{Module,Array{Any,1},Int64})() at /home/rik/git/FR/src/eval/run.jl:65
```

and ignore everything that comes afterwards which is just noise

```
r = r"\[\d+\]\s\(\:\:Franklin.var(?:.*?)src\/eval\/run.jl"
```

should do the trick, a line that has a match for that can be removed as well as all lines below it, what do you think?

**Comment by rikhuijzer** [comment]
I knew that I was forgetting something! Yes, you mentioned that. Good idea :+1:  I'll add it somewhere today

**Comment by rikhuijzer** [comment]
The Franklin Warning is now
```
┌ Franklin Warning: in <input string>
│ There was an error of type 'DomainError' when running a code block.
│ Checking the output files '/tmp/jl_meCJZ8.(out|res)'
│ might be helpful to understand and solve the issue.
│
│ Relevant pointers:
│ - evaluation of Julia code blocks: https://franklinjl.org/code/
│
│ Stacktrace:
│ DomainError with -1.0:
│ sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
│ Stacktrace:
│  [1] throw_complex_domainerror(::Symbol, ::Float64) at ./math.jl:33
│  [2] sqrt at ./math.jl:573 [inlined]
│  [3] sqrt(::Int64) at ./math.jl:599
│  [4] top-level scope at none:1
│  [5] eval at ./boot.jl:331 [inlined]
└
```

File: src/eval/run.jl [hunk 1]
```diff
@@ -75,15 +80,20 @@ function run_code(mod::Module, code::AS, out_path::AS;
     end
     # if there was an error, return nothing and possibly show warning
     if !isnothing(err)
-        # TODO: add more informative message, maybe show type of error
-        # + parent path
+        function strip_franklin(stacktrace::String)
+            rx = r"\[\d+\]\s\(\:\:Franklin.var(?:.*?)src\/eval\/run.jl"
+            first_match_start = first(first(findall(rx, stacktrace)))
+            stacktrace[1:first_match_start-3]
+        end
```
Comment (tlienart) [inline src/eval/run.jl:87]: hmm that's a bit of an ugly spot to put that function; maybe just put it outside, call it `trim_stacktrace`, and use your previous stack trace as a test?

**Comment by rikhuijzer** [comment]
Good suggestion; I've added it. Also, "stacktrace" was in the output twice, so I changed it to "details".

File: src/eval/run.jl [hunk 2]
```diff
@@ -23,6 +23,12 @@ function parse_code(code::AS)
     exs
 end
 
+function trim_stacktrace(stacktrace::String)
```
Comment (tlienart) [inline src/eval/run.jl:26]: Could you just add a very short doc-string clarifying that you're cleaning up the stack trace beyond things that are unrelated to the actual error?

File: src/eval/run.jl [hunk 3]
```diff
@@ -23,6 +23,12 @@ function parse_code(code::AS)
...
+    rx = r"\[\d+\]\s\(\:\:Franklin.var(?:.*?)src\/eval\/run.jl"
+    first_match_start = first(first(findall(rx, stacktrace)))
```
Comment (tlienart) [inline src/eval/run.jl:28]: here you can use `findfirst` I think

File: src/eval/run.jl [hunk 4]
```diff
@@ -23,6 +23,12 @@ function parse_code(code::AS)
...
+    stacktrace[1:first_match_start-3]
```
Comment (tlienart) [inline src/eval/run.jl:29]: might be silly but just add a comment explaining what you're doing here; this is just so that the future me (or whomever else) who comes back to this doesn't  have to  scratch  their head for too long :joy:

**Review by tlienart — COMMENTED** [review]
Some very small comments and then I'll merge; thanks  again this is great!

**Comment by rikhuijzer** [comment]
No problem. Rome wasn't built in a day. 

I've applied the comments.

**Comment by tlienart** [comment]
Thanks so much!
